README.md
## iOS 12 & Swift - The Complete iOS App Development Bootcamp

### Machine Learning

- Integração de modelo de reconhecimeto de imagens de plantas em um aplicativo escrito em swift para iOS.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added
### Changed
### Removed
### Fixed

## [0.1.1] 2019-06-30
### Added
- Adicionado pod SwiftyJson.
- Adicionado pod Alamofire.
- Adicionado metodo get para recuperar informação da classificação.
- Adicionado label para descrição da imagem classificada.
- Ajustes do layout de tela.
- Parse Json.
- Adicionado pod para download de imagem.
- Adicionado imagem de classificação da wikipedia.


## [0.1.0] 2019-06-29
### Added
- padrão MVC.
- Adicionado modelo caffe model.
- Adicionado navigation controller.
- Adicionado metodo para tirar foto.
- Adicionado metedo para classificação de imagem.


WhatPlant/AppDelegate.swift
//
//  AppDelegate.swift
//  WhatPlant
//
//  Created by Maurício de Freitas Sayão on 29/06/19.
//  Copyright © 2019 Maurício de Freitas Sayão. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        return true
    }

    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and invalidate graphics rendering callbacks. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the active state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



WhatPlant/control/ViewController.swift
//
//  ViewController.swift
//  WhatPlant
//
//  Created by Maurício de Freitas Sayão on 29/06/19.
//  Copyright © 2019 Maurício de Freitas Sayão. All rights reserved.
//

import UIKit
import CoreML
import Vision
import SwiftyJSON
import Alamofire
import SDWebImage

class ViewController: UIViewController {
    
    @IBOutlet weak var imageView: UIImageView!
    @IBOutlet weak var flowerInfoLabel: UILabel!
    var imagePicker = UIImagePickerController()
    let wikipidiaUrl = "https://en.wikipedia.org/w/api.php"
    
    override func viewDidLoad() {
        super.viewDidLoad()
        imagePicker.delegate = self
    }
    
    @IBAction func buttonTapped(_ sender: UIBarButtonItem) {
        
        imagePicker.allowsEditing = true
        imagePicker.sourceType = .camera
        present(imagePicker, animated: true, completion: nil)
        
    }
    
    private func detect(image: CIImage) {
        guard let model = try? VNCoreMLModel(for: FlowersClassifiers().model) else {
            fatalError("Load model failed.")
        }
        
        let request = VNCoreMLRequest(model: model) { (req, error) in
            if error != nil {
                print(error.debugDescription)
                return
            }
            
            guard let reqResults = req.results as? [VNClassificationObservation] else {
                fatalError("Error on Classifier")
            }
            
            guard let first = reqResults.first  else {
                fatalError("Error on Classifier image")
            }
            
            self.getInfo(flowerName: first.identifier)
        }
        
        let handler = VNImageRequestHandler(ciImage: image)
        
        do {
            try handler.perform([request])
        }catch{
            print(error)
        }
    }
    
    private func getInfo(flowerName: String) {
        
        let parameters : [String: String] = [
            "format": "json",
            "action": "query",
            "prop": "extracts|pageimages",
            "exintro": "",
            "explaintext": "",
            "titles": flowerName,
            "indexpageids": "",
            "redirects": "1",
            "pithumbsize": "500"
        ]
        
        Alamofire.request(wikipidiaUrl, method: .get, parameters: parameters).responseJSON { (response) in
            if response.result.isSuccess {
                let flower = self.toObject(json: JSON(response.result.value!))
                
                self.imageView.sd_setImage(with: URL(string: flower.imageUrl), completed: nil)
                self.navigationItem.title = flower.name
                self.flowerInfoLabel.text = flower.description
                
            }
        }
    }
    
    private func toObject(json: JSON) -> Flower {
        let pageId = json["query"]["pageids"][0].stringValue
        
        let flowerName = json["query"]["pages"][pageId]["title"].stringValue
        let flowerInfo = json["query"]["pages"][pageId]["extract"].stringValue
        let flowerImg = json["query"]["pages"][pageId]["thumbnail"]["source"].stringValue
        
        return Flower(name: flowerName, description: flowerInfo, imageUrl: flowerImg)
    }
    
}

extension ViewController: UIImagePickerControllerDelegate {
    
    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) {
        
        if let image = info[UIImagePickerController.InfoKey.editedImage] as? UIImage {
            guard let imageEdited = CIImage(image: image) else {
                fatalError("Error on convert image!")
            }
            
            detect(image: imageEdited)
        }
        imagePicker.dismiss(animated: true, completion: nil)
    }
    
}

extension ViewController: UINavigationControllerDelegate {
    
}

